# HG changeset patch
# User alta88@fixall.com
# Date 1692729444 -10800
# Parent  a78ce60b8b653382d8c1817d08cd5ce390f2ef1f
Attachment List on top option for messages
Implement a collapsible message headers splitter/sizer
- keyboard support for sizing splitters with arrow keys, toggling with enter/double-click
* * *
Optimize height setting for messages attachment view.
* * *

Authored by Alta88 and contributed to the Betterbird Project.
See https://github.com/Betterbird/thunderbird-patches/blob/main/LICENSE for license details.
All uses require attribution to the Author.

diff --git a/calendar/base/themes/common/calendar.css b/calendar/base/themes/common/calendar.css
--- a/calendar/base/themes/common/calendar.css
+++ b/calendar/base/themes/common/calendar.css
@@ -124,17 +124,17 @@
   }
 
   .calendar-notification-bar:not(:-moz-lwtheme) > *:not(#imip-view-toolbox) {
     color: var(--selected-item-text-color);
   }
 }
 
 .calendar-notification-bar {
-  margin: 0 4px 4px;
+  margin: 4px;
   border-radius: 4px;
   box-shadow: 0 1px 2px rgba(58, 57, 68, 0.3);
 }
 
 #imip-view-toolbar {
   --imip-button-background: var(--layout-background-1);
 }
 
diff --git a/mail/app/profile/all-thunderbird.js b/mail/app/profile/all-thunderbird.js
--- a/mail/app/profile/all-thunderbird.js
+++ b/mail/app/profile/all-thunderbird.js
@@ -326,16 +326,17 @@ pref("mail.ui-rdf.version", 0);
 // Overrides of the core mailnews.js and composer.js prefs
 /////////////////////////////////////////////////////////////////
 pref("mail.showCondensedAddresses", true); // show the friendly display name for people I know
 // 1: Show display name, 2: Show e-mail address, 3: Show both.
 pref("mail.addressDisplayAuthor", 1);
 pref("mail.addressDisplayRecipients", 1);
 
 pref("mailnews.attachments.display.start_expanded", false);
+pref("mailnews.attachments.display.top", false);
 // hidden pref for changing how we present attachments in the message pane
 pref("mail.pane_config.dynamic",            0);
 pref("mail.pane_config.multiline_all", false);
 pref("mailnews.reuse_thread_window2",     true);
 pref("editor.singleLine.pasteNewlines", 4);  // substitute commas for new lines in single line text boxes
 pref("editor.CR_creates_new_p", true);
 pref("mail.compose.default_to_paragraph", true);
 
diff --git a/mail/base/content/aboutMessage.js b/mail/base/content/aboutMessage.js
--- a/mail/base/content/aboutMessage.js
+++ b/mail/base/content/aboutMessage.js
@@ -9,18 +9,19 @@
      nsMsgViewIndex_None, TreeSelection */
 /* globals gDBView: true, gFolder: true, gViewWrapper: true */
 
 // mailContext.js
 /* globals mailContextMenu */
 
 // msgHdrView.js
 /* globals AdjustHeaderView ClearCurrentHeaders ClearPendingReadTimer
-   HideMessageHeaderPane OnLoadMsgHeaderPane OnTagsChange
-   OnUnloadMsgHeaderPane HandleAllAttachments AttachmentMenuController */
+   gHeaderCustomize HideMessageHeaderPane OnLoadMsgHeaderPane OnTagsChange
+   OnUnloadMsgHeaderPane HandleAllAttachments AttachmentMenuController
+   SetAttachmentListLocation */
 
 var { MailServices } = ChromeUtils.import(
   "resource:///modules/MailServices.jsm"
 );
 var { XPCOMUtils } = ChromeUtils.importESModule(
   "resource://gre/modules/XPCOMUtils.sys.mjs"
 );
 
@@ -53,16 +54,156 @@ function messagePaneOnResize() {
     img.toggleAttribute(
       "overflowing",
       img.clientWidth - doc.body.offsetWidth >= 0 &&
         (img.clientWidth <= img.naturalWidth || !img.naturalWidth)
     );
   }
 }
 
+function onHeaderSplitterAction(event) {
+  onSplitterAction(event);
+  if (
+    event.keyCode != KeyEvent.DOM_VK_RETURN &&
+    event.type != "dblclick" &&
+    event.type != "mouseup"
+  ) {
+    return;
+  }
+  let splitter = event.target;
+  let resizeElement = document.getElementById(
+    splitter.getAttribute("resize-id")
+  );
+  let messageHeader = document.getElementById("messageHeader");
+  let isCollapsed = splitter.getAttribute("state") == "collapsed";
+
+  if (gHeaderCustomize.customizeData.collapsed != isCollapsed) {
+    gHeaderCustomize.customizeData.collapsed = isCollapsed;
+    gHeaderCustomize.onPanelHidden();
+  }
+
+  if (
+    event.type == "mouseup" &&
+    !isCollapsed &&
+    messageHeader.clientHeight < resizeElement.getAttribute("height")
+  ) {
+    // Necessary to wait for reflow to happen on splitter resize.
+    setTimeout(() => {
+      resizeElement.style.height = "";
+      resizeElement.setAttribute("height", messageHeader.clientHeight);
+    });
+  }
+}
+
+function onAttachmentSplitterAction(event) {
+  onSplitterAction(event);
+  if (event.type != "mouseup") {
+    return;
+  }
+  let splitter = event.target;
+  let resizeElement = document.getElementById(
+    splitter.getAttribute("resize-id")
+  );
+  let attachmentList = document.getElementById("attachmentList");
+  let attachmentBar = document.getElementById("attachmentBar");
+  let optimalHeight =
+    attachmentBar.clientHeight + attachmentList.scrollHeight;
+  let isCollapsed = splitter.getAttribute("state") == "collapsed";
+
+  if (!isCollapsed && optimalHeight < resizeElement.getAttribute("height")) {
+    // Necessary to wait for reflow to happen on splitter resize.
+    setTimeout(() => {
+      resizeElement.style.height = "";
+      resizeElement.removeAttribute("height");
+    });
+  }
+}
+
+/*
+ * Click or keypress. Expand/collapse with dblclick or enter; resize with arrow
+ * keys.
+ */
+function onSplitterAction(event) {
+  if (event.ctrlKey || event.altKey || event.metaKey || event.shiftKey) {
+    return;
+  }
+
+  let splitter = event.target;
+  let resizeElement = document.getElementById(
+    splitter.getAttribute("resize-id")
+  );
+
+  if (!resizeElement) {
+    return;
+  }
+
+  let isVertical = splitter.getAttribute("orient") == "vertical";
+  let dimension = isVertical ? "height" : "width";
+  let isCollapsed = splitter.getAttribute("state") == "collapsed";
+  let minDefaultSize = isVertical ? 30 : 100;
+
+  if (event.keyCode == KeyEvent.DOM_VK_RETURN || event.type == "dblclick") {
+    if (isCollapsed) {
+      if (resizeElement.getBoundingClientRect()[dimension] == 0) {
+        let collapseSize =
+          Number(splitter.getAttribute(`collapse-${dimension}`)) ||
+          minDefaultSize;
+        let size = Number(resizeElement.getAttribute(dimension));
+        resizeElement.style[dimension] = (size || collapseSize) + "px";
+      }
+    }
+    splitter.setAttribute("state", isCollapsed ? "" : "collapsed");
+    if (event.keyCode == KeyEvent.DOM_VK_RETURN) {
+      return;
+    }
+  }
+
+  if (event.type == "dblclick" || event.type == "mouseup") {
+    splitter.blur();
+    return;
+  }
+
+  let isUpDownKey = [KeyEvent.DOM_VK_UP, KeyEvent.DOM_VK_DOWN].includes(
+    event.keyCode
+  );
+  let isLeftRightKey = [KeyEvent.DOM_VK_LEFT, KeyEvent.DOM_VK_RIGHT].includes(
+    event.keyCode
+  );
+  let isUpLeftKey = [KeyEvent.DOM_VK_UP, KeyEvent.DOM_VK_LEFT].includes(
+    event.keyCode
+  );
+  let isDownRightKey = [KeyEvent.DOM_VK_DOWN, KeyEvent.DOM_VK_RIGHT].includes(
+    event.keyCode
+  );
+  let isValidKey = isUpDownKey || isLeftRightKey;
+  if (
+    !isValidKey ||
+    (isVertical && !isUpDownKey) ||
+    (!isVertical && !isLeftRightKey)
+  ) {
+    return;
+  }
+
+  let _beforeElement = !(
+    splitter.compareDocumentPosition(resizeElement) &
+    Node.DOCUMENT_POSITION_FOLLOWING
+  );
+  let size = isCollapsed
+    ? 0
+    : Math.round(resizeElement.getBoundingClientRect()[dimension]);
+  let newSize =
+    (isUpLeftKey && _beforeElement) || (isDownRightKey && !_beforeElement)
+      ? size - 1
+      : size + 1;
+  newSize = newSize < 0 ? 0 : newSize;
+  resizeElement.style[dimension] = newSize + "px";
+  resizeElement.setAttribute(dimension, newSize);
+  splitter.setAttribute("state", newSize > 0 ? "" : "collapsed");
+}
+
 function ReloadMessage() {
   if (!gMessageURI) {
     return;
   }
   displayMessage(gMessageURI, gViewWrapper);
 }
 
 async function MailSetCharacterSet() {
@@ -160,16 +301,18 @@ window.addEventListener("DOMContentLoade
   Enigmail.msg.messengerStartup();
   Enigmail.hdrView.hdrViewLoad();
 
   MailServices.mailSession.AddFolderListener(
     folderListener,
     Ci.nsIFolderListener.removed
   );
 
+  SetAttachmentListLocation();
+
   preferenceObserver.init();
   Services.obs.addObserver(msgObserver, "message-content-updated");
 
   if (parent == top) {
     // Standalone message display? Focus the message pane.
     getMessagePaneBrowser().focus();
   }
 
@@ -354,16 +497,17 @@ var msgObserver = {
 
 var preferenceObserver = {
   QueryInterface: ChromeUtils.generateQI(["nsIObserver"]),
 
   _topics: [
     "mail.inline_attachments",
     "mail.show_headers",
     "mail.showCondensedAddresses",
+    "mailnews.attachments.display.top",
     "mailnews.display.disallow_mime_handlers",
     "mailnews.display.html_as",
     "mailnews.display.prefer_plaintext",
     "mailnews.headers.showReferences",
     "rss.show.summary",
   ],
 
   _reloadTimeout: null,
@@ -379,16 +523,20 @@ var preferenceObserver = {
       Services.prefs.removeObserver(topic, this);
     }
   },
 
   observe(subject, topic, data) {
     if (data == "mail.show_headers") {
       AdjustHeaderView(Services.prefs.getIntPref(data));
     }
+    if (data == "mailnews.attachments.display.top") {
+      SetAttachmentListLocation();
+      return;
+    }
     if (!this._reloadTimeout) {
       // Clear the event queue before reloading the message. Several prefs may
       // be changed at once.
       this._reloadTimeout = setTimeout(() => {
         this._reloadTimeout = null;
         ReloadMessage();
       });
     }
diff --git a/mail/base/content/aboutMessage.xhtml b/mail/base/content/aboutMessage.xhtml
--- a/mail/base/content/aboutMessage.xhtml
+++ b/mail/base/content/aboutMessage.xhtml
@@ -102,18 +102,25 @@
     <!-- msg header view -->
     <!-- a convenience box for ease of extension overlaying -->
     <hbox id="messagepaneboxwrapper" flex="1">
       <vbox id="messagepanebox">
         <vbox id="singleMessage">
           <hbox id="msgHeaderView" collapsed="true" class="main-header-area">
 #include msgHdrView.inc.xhtml
           </hbox>
+        </vbox>
+        <splitter id="header-splitter" orient="vertical"
+                  resizebefore="closest" resizeafter="none"
+                  collapse="before" collapsed="true"
+                  resize-id="singleMessage" tabindex="0"
+                  onkeypress="onHeaderSplitterAction(event);"
+                  ondblclick="onHeaderSplitterAction(event)"
+                  onmouseup="onHeaderSplitterAction(event)"/>
 #include ../../../calendar/base/content/imip-bar-overlay.inc.xhtml
-        </vbox>
         <!-- The msgNotificationBar appears on top of the message and displays
              information like: junk, mdn, remote content and phishing warnings -->
         <vbox id="mail-notification-top">
           <!-- notificationbox will be added here lazily. -->
         </vbox>
 
 #include ../../../calendar/base/content/widgets/calendar-invitation-panel.xhtml
 #include ../../../calendar/base/content/widgets/calendar-minidate.xhtml
@@ -125,34 +132,38 @@
            <!-- The calendar invitation panel is displayed here. -->
           </html:div>
         </vbox>
 
         <!-- message view -->
         <browser id="messagepane"
                  context="mailContext"
                  tooltip="aHTMLTooltip"
-                 style="height: 0px; min-height: 1px; background-color: field;"
+                 style="height: 0px; min-height: 10px; background-color: field;"
                  flex="1"
                  name="messagepane"
                  disablesecurity="true"
                  disablehistory="true"
                  type="content"
                  primary="true"
                  autofind="false"
                  nodefaultsrc="true"
                  forcemessagemanager="true"
                  maychangeremoteness="true"
                  messagemanagergroup="single-page"
                  onclick="return contentAreaClick(event);"
                  onresize="messagePaneOnResize();"/>
-        <splitter id="attachment-splitter" orient="vertical"
-                  resizebefore="closest" resizeafter="closest"
-                  collapse="after" collapsed="true"/>
-        <vbox id="attachmentView" collapsed="true">
+        <splitter id="attachment-splitter" location="bottom" orient="vertical"
+                  resizebefore="none" resizeafter="closest"
+                  collapse="after" collapsed="true"
+                  resize-id="attachmentView" tabindex="0"
+                  onkeypress="onAttachmentSplitterAction(event)"
+                  ondblclick="onAttachmentSplitterAction(event)"
+                  onmouseup="onAttachmentSplitterAction(event)"/>
+        <vbox id="attachmentView" location="bottom" collapsed="true">
 #include msgAttachmentView.inc.xhtml
         </vbox>
         <findbar id="FindToolbar" browserid="messagepane"/>
       </vbox>
 #include msgSecurityPane.inc.xhtml
     </hbox>
 </html:body>
 </html>
diff --git a/mail/base/content/msgHdrView.js b/mail/base/content/msgHdrView.js
--- a/mail/base/content/msgHdrView.js
+++ b/mail/base/content/msgHdrView.js
@@ -1232,25 +1232,16 @@ function RemoveNewHeaderViews(aHeaderTab
  * and see if we have a corresponding entry for that header (i.e.
  * whether the expanded header view cares about this header value)
  * If so, then call updateHeaderEntry
  */
 function UpdateExpandedMessageHeaders() {
   // Iterate over each header we received and see if we have a matching entry
   // in each header view table...
   var headerName;
-
-  // Remove the height attr so that it redraws correctly. Works around a problem
-  // that attachment-splitter causes if it's moved high enough to affect
-  // the header box:
-  document.getElementById("msgHeaderView").removeAttribute("height");
-  // This height attribute may be set by toggleWrap() if the user clicked
-  // the "more" button" in the header.
-  // Remove it so that the height is determined automatically.
-
   for (headerName in currentHeaderData) {
     var headerField = currentHeaderData[headerName];
     var headerEntry = null;
 
     if (headerName in gExpandedHeaderView) {
       headerEntry = gExpandedHeaderView[headerName];
     }
 
@@ -1344,27 +1335,29 @@ function ClearCurrentHeaders() {
   currentHeaderData = {};
   // eslint-disable-next-line no-global-assign
   currentAttachments = [];
   currentCharacterSet = "";
 }
 
 function ShowMessageHeaderPane() {
   document.getElementById("msgHeaderView").collapsed = false;
+  document.getElementById("header-splitter").collapsed = false;
   document.getElementById("mail-notification-top").collapsed = false;
 
   // Initialize the DBListener if we don't have one. This might happen when the
   // message pane is hidden or no message was selected before, which caused the
   // clearing of the the DBListener.
   initFolderDBListener();
 }
 
 function HideMessageHeaderPane() {
   let header = document.getElementById("msgHeaderView");
   header.collapsed = true;
+  document.getElementById("header-splitter").collapsed = true;
   document.getElementById("mail-notification-top").collapsed = true;
 
   // Disable the attachment box.
   document.getElementById("attachmentView").collapsed = true;
   document.getElementById("attachment-splitter").collapsed = true;
 
   gMessageNotificationBar.clearMsgNotifications();
   // Clear the DBListener since we don't have any visible UI to update.
@@ -1453,16 +1446,83 @@ function outputEmailAddresses(headerEntr
     address.displayName = addr.name;
     headerEntry.enclosingBox.addRecipient(address);
   }
 
   headerEntry.enclosingBox.buildView();
 }
 
 /**
+ * Set the location of the attachment list toolbar, per pref.
+ */
+function SetAttachmentListLocation() {
+  let pref = Services.prefs.getBoolPref("mailnews.attachments.display.top");
+  let location = pref ? "top" : "bottom";
+  let attachmentView = document.getElementById("attachmentView");
+  if (attachmentView.getAttribute("location") == location) {
+    // Already in the right location.
+    return;
+  }
+  const attachmentListConfigMap = {
+    top: {
+      parentId: "messagepanebox",
+      insertbeforeId: "imip-bar",
+      splitterOrient: "vertical",
+      splitterCollapse: "before",
+    },
+    bottom: {
+      parentId: "messagepanebox",
+      insertbeforeId: "FindToolbar",
+      splitterOrient: "vertical",
+      splitterCollapse: "after",
+    },
+  };
+
+  let configMap = attachmentListConfigMap[location];
+  let attachmentSplitter = document.getElementById("attachment-splitter");
+  let attachmentList = document.getElementById("attachmentList");
+
+  let parentId = configMap.parentId;
+  let insertBeforeId = configMap.insertbeforeId;
+  let splitterOrient = configMap.splitterOrient;
+  let splitterCollapse = configMap.splitterCollapse;
+
+  let desiredParent = document.getElementById(parentId);
+  let insertBefore = document.getElementById(insertBeforeId);
+
+  switch (location) {
+    case "top":
+      desiredParent.insertBefore(attachmentView, insertBefore);
+      desiredParent.insertBefore(attachmentSplitter, insertBefore);
+      attachmentSplitter.setAttribute("resizebefore", "closest");
+      attachmentSplitter.setAttribute("resizeafter", "none");
+      break;
+    case "bottom":
+      desiredParent.insertBefore(attachmentSplitter, insertBefore);
+      desiredParent.insertBefore(attachmentView, insertBefore);
+      attachmentSplitter.setAttribute("resizebefore", "none");
+      attachmentSplitter.setAttribute("resizeafter", "closest");
+      break;
+  }
+
+  document.getElementById("imip-bar").style.height = "";
+  document.getElementById("mail-notification-top").style.height = "";
+
+  attachmentView.setAttribute("location", location);
+
+  attachmentList.setAttribute(
+    "orient",
+    splitterOrient == "horizontal" ? "vertical" : "horizontal"
+  );
+  attachmentSplitter.setAttribute("orient", splitterOrient);
+  attachmentSplitter.setAttribute("collapse", splitterCollapse);
+  attachmentSplitter.setAttribute("location", location);
+}
+
+/**
  * Return true if possible attachments in the currently loaded message can be
  * deleted/detached.
  */
 function CanDetachAttachments() {
   var canDetach =
     !gFolder.isSpecialFolder(Ci.nsMsgFolderFlags.Newsgroup, false) &&
     (!gFolder.isSpecialFolder(Ci.nsMsgFolderFlags.ImapBox, false) ||
       MailOfflineMgr.isOnline()) &&
@@ -2075,38 +2135,34 @@ function toggleAttachmentList(expanded, 
     }
     attachmentBar.setAttribute(
       "tooltiptext",
       bundle.getString("collapseAttachmentPaneTooltip")
     );
 
     attachmentList.setOptimumWidth();
 
-    // By design, attachmentView should not take up more than 1/4 of the message
-    // pane space
-    attachmentView.setAttribute(
-      "height",
-      Math.min(
-        attachmentList.preferredHeight,
-        document.getElementById("messagepanebox").getBoundingClientRect()
-          .height / 4
-      )
-    );
+    // Set either the user height or the optimal height (if smaller). The
+    // view is limited to max-height: 50%.
+    let preferredHeight = attachmentView.getAttribute("height");
+    let optimalHeight =
+      attachmentBar.clientHeight + attachmentList.scrollHeight;
+    attachmentView.style.height = (preferredHeight || optimalHeight) + "px";
 
     if (updateFocus) {
       attachmentList.focus();
     }
   } else {
     attachmentList.collapsed = true;
     attachmentSplitter.collapsed = true;
     attachmentBar.setAttribute(
       "tooltiptext",
       bundle.getString("expandAttachmentPaneTooltip")
     );
-    attachmentView.removeAttribute("height");
+    attachmentView.style.height = "";
 
     if (updateFocus && document.activeElement == attachmentList) {
       // TODO
     }
   }
 }
 
 /**
@@ -2534,24 +2590,26 @@ const gHeaderCustomize = {
    * @property {boolean} hideLabels - If the labels column should be hidden.
    * @property {boolean} subjectLarge - If the font size of the subject line
    *   should be increased.
    * @property {string} buttonStyle - The style in which the buttons should be
    *   rendered:
    *   - "default" = icons+text
    *   - "only-icons" = only icons
    *   - "only-text" = only text
+   * @property {boolean} collapsed - Hide headers pane if true.
    */
   customizeData: {
     showAvatar: true,
     showBigAvatar: false,
     showFullAddress: true,
     hideLabels: true,
     subjectLarge: true,
     buttonStyle: "default",
+    collapsed: false,
   },
 
   /**
    * Initialize the customizer.
    */
   init() {
     this.customizePanel = document.getElementById(
       "messageHeaderCustomizationPanel"
@@ -2565,16 +2623,17 @@ const gHeaderCustomize = {
     }
   },
 
   /**
    * Reset and update the customized style of the message header.
    */
   updateLayout() {
     let header = document.getElementById("messageHeader");
+    let headerSplitter = document.getElementById("header-splitter");
     // Always clear existing styles to avoid visual issues.
     header.classList.remove(
       "message-header-large-subject",
       "message-header-buttons-only-icons",
       "message-header-buttons-only-text",
       "message-header-hide-label-column"
     );
 
@@ -2626,16 +2685,21 @@ const gHeaderCustomize = {
       default:
         header.classList.remove(
           "message-header-buttons-only-icons",
           "message-header-buttons-only-text"
         );
         break;
     }
 
+    headerSplitter.setAttribute(
+      "state",
+      this.customizeData.collapsed ? "collapsed" : ""
+    );
+
     gMessageHeader.syncLabelsColumnWidths();
   },
 
   /**
    * Show the customization panel for the message header.
    */
   showPanel() {
     this.customizePanel.openPopup(
diff --git a/mail/base/content/widgets/mailWidgets.js b/mail/base/content/widgets/mailWidgets.js
--- a/mail/base/content/widgets/mailWidgets.js
+++ b/mail/base/content/widgets/mailWidgets.js
@@ -626,25 +626,16 @@
           child.setAttribute("context", this.getAttribute("itemcontext"))
         );
     }
 
     get itemCount() {
       return this._childNodes.length;
     }
 
-    /**
-     * Get the preferred height (the height that would allow us to fit
-     * everything without scrollbars) of the attachmentlist's bounding
-     * rectangle. Add 3px to account for item's margin.
-     */
-    get preferredHeight() {
-      return this.scrollHeight + this.getBoundingClientRect().height + 3;
-    }
-
     get _childNodes() {
       return this.querySelectorAll("richlistitem.attachmentItem");
     }
 
     getIndexOfItem(item) {
       for (let i = 0; i < this._childNodes.length; i++) {
         if (this._childNodes[i] === item) {
           return i;
diff --git a/mail/components/preferences/general.inc.xhtml b/mail/components/preferences/general.inc.xhtml
--- a/mail/components/preferences/general.inc.xhtml
+++ b/mail/components/preferences/general.inc.xhtml
@@ -149,16 +149,24 @@
       <hbox>
         <vbox>
           <checkbox id="verticalTabs"
                     data-l10n-id="vertical-tabs-label"
                     preference="mail.tabs.vertical"/>
         </vbox>
         <spacer flex="1"/>
       </hbox>
+      <hbox>
+        <vbox>
+          <checkbox id="attachmentListOnTopMessages"
+                    data-l10n-id="attachments-display-top-label"
+                    preference="mailnews.attachments.display.top"/>
+        </vbox>
+        <spacer flex="1"/>
+      </hbox>
     </html:fieldset>
     </html:div>
 
     <!-- Fonts and Colors -->
     <html:div data-category="paneGeneral">
     <html:fieldset id="fontsGroup" data-category="paneGeneral">
       <html:legend data-l10n-id="fonts-legend"></html:legend>
 
diff --git a/mail/components/preferences/general.js b/mail/components/preferences/general.js
--- a/mail/components/preferences/general.js
+++ b/mail/components/preferences/general.js
@@ -89,16 +89,17 @@ Preferences.addAll([
   { id: "browser.cache.disk.capacity", type: "int" },
   { id: "browser.cache.disk.smart_size.enabled", inverted: true, type: "bool" },
   { id: "privacy.clearOnShutdown.cache", type: "bool" },
   { id: "layers.acceleration.disabled", type: "bool", inverted: true },
   { id: "searchintegration.enable", type: "bool" },
   { id: "mail.tabs.drawInTitlebar", type: "bool" },
   { id: "mail.tabs.autoHide", type: "bool" },
   { id: "mail.tabs.vertical", type: "bool" },
+  { id: "mailnews.attachments.display.top", type: "bool" },
 ]);
 if (AppConstants.platform == "win") {
   Preferences.add({ id: "mail.minimizeToTray", type: "bool" });
 }
 if (AppConstants.platform != "macosx") {
   Preferences.add({ id: "mail.biff.show_alert", type: "bool" });
 }
 
diff --git a/mail/locales/en-US/messenger/preferences/preferences.ftl b/mail/locales/en-US/messenger/preferences/preferences.ftl
--- a/mail/locales/en-US/messenger/preferences/preferences.ftl
+++ b/mail/locales/en-US/messenger/preferences/preferences.ftl
@@ -276,16 +276,20 @@ auto-hide-tabbar-label =
     .label = Auto hide tab bar
     .accesskey = A
 auto-hide-tabbar-description = Hide the tab bar when only a single tab is open
 
 vertical-tabs-label =
     .label = Enable vertical tabs
     .accesskey = v
 
+attachments-display-top-label =
+    .label = Display attachment list above message
+    .accesskey = p
+
 system-integration-legend = System Integration
 always-check-default =
     .label = Always check to see if { -brand-short-name } is the default mail client on startup
     .accesskey = A
 check-default-button =
     .label = Check Now…
     .accesskey = N
 
diff --git a/mail/themes/shared/mail/attachmentList.css b/mail/themes/shared/mail/attachmentList.css
--- a/mail/themes/shared/mail/attachmentList.css
+++ b/mail/themes/shared/mail/attachmentList.css
@@ -9,17 +9,16 @@
   padding: 3px;
   background-color: var(--layout-background-2);
   color: var(--layout-color-2);
   border: none;
   display: flex;
   overflow-x: hidden;
   align-items: start;
   align-content: start;
-  flex: 1;
   flex-direction: row;
   flex-wrap: wrap;
 }
 
 .attachmentList[collapsed] {
   display: none;
   height: 0;
 }
diff --git a/mail/themes/shared/mail/messageHeader.css b/mail/themes/shared/mail/messageHeader.css
--- a/mail/themes/shared/mail/messageHeader.css
+++ b/mail/themes/shared/mail/messageHeader.css
@@ -51,19 +51,29 @@
   }
 }
 
 #messagepanebox {
   color: var(--layout-color-1);
   background-color: var(--layout-background-1);
 }
 
+/* Protect against the headers overflowing the entire preview. */
+#singleMessage {
+  max-height: 85%;
+}
+
 .main-header-area {
   border-bottom-style: none;
   display: block;
+  overflow-x: clip;
+  overflow-y: auto;
+  /* We need the following to lose the scroll bar when making this high
+   * enough via the splitter */
+  height: 100%;
 }
 
 .message-header-container,
 .message-header-extra-container {
   display: grid;
   row-gap: 6px;
 }
 
@@ -104,31 +114,34 @@
 }
 
 .header-row-grow:not([hidden]) {
   flex: 1 1 auto;
   display: flex;
   align-items: center;
 }
 
-#mail-notification-top {
+#mail-notification-top > .notificationbox-stack:not(:empty) {
   border-bottom: 1px solid var(--splitter-color);
 }
 
 #mail-notification-top > .notificationbox-stack[notificationside="top"] {
   background-color: var(--layout-background-1);
 }
 
 /* ::::: msg header toolbars ::::: */
 
-#messageHeader[show_header_mode="all"],
-#messageHeader.scrollable {
-  overflow-y: auto;
-  overflow-x: hidden;
-  max-height: 14em;
+#messageHeader[show_header_mode="all"] {
+  max-height: 18em;
+}
+
+/* Height restriction for scrollable gets in the way of dragging it as
+ * big as we want. */
+#messageHeader.scrollable:not([show_header_mode="all"]) {
+  overflow: hidden;
 }
 
 #expandedBoxSpacer {
   display: block;
   height: 4px;
 }
 
 mail-tagfield[collapsed="true"] {
@@ -264,27 +277,63 @@ mail-tagfield[collapsed="true"] {
   padding-inline: var(--message-header-field-offset);
   /* IMPORTANT! Keep these to avoid issues with very long subjects. Bug 77806 */
   display: -webkit-box;
   -webkit-box-orient: vertical;
   -webkit-line-clamp: 3;
   overflow-y: auto;
 }
 
+/* ::::: Splitters, header and attachments ::::: */
+
+splitter[orient="vertical"] {
+  appearance: none;
+  background-color: transparent;
+  border-width: 0;
+  border-top: 1px solid var(--splitter-color);
+  /* splitter grip area */
+  height: 5px;
+  /* make only the splitter border visible */
+  margin-bottom: -5px;
+  /* because of the negative margin needed to make the splitter visible */
+  position: relative;
+  z-index: 10;
+}
+
+@media (prefers-reduced-motion: no-preference) {
+  splitter[orient="vertical"] {
+    transition: border-color 200ms;
+  }
+}
+
+splitter[orient="vertical"]:hover,
+splitter[orient="vertical"]:focus {
+  border-top: 4px solid var(--selected-item-color);
+}
+
+splitter[state="collapsed"]:not(:hover, :focus) {
+  border-top-width: 0;
+}
+
 /* ::::: attachment view ::::: */
 
 #attachmentBar {
   display: grid;
   grid-template-columns: max-content auto max-content;
   align-items: center;
   color: var(--layout-color-1);
   background-color: var(--layout-background-1);
   padding: 3px 0;
 }
 
+/* Collapsed list */
+#attachmentView[location="top"] #attachmentBar[tooltiptext^="Show"] {
+  border-bottom: 1px solid var(--splitter-color);
+}
+
 #attachmentToggle {
   /* Override button appearance */
   appearance: none;
   min-width: 20px;
   margin-block: 0;
   margin-inline: 1px 0;
   border: 0;
   background-color: transparent;
@@ -358,41 +407,28 @@ mail-tagfield[collapsed="true"] {
   border-radius: 2px;
 }
 
 #attachmentName:focus-visible {
   outline: var(--focus-outline);
   outline-offset: -1px;
 }
 
-#attachment-splitter {
-  appearance: none;
-  background-color: transparent;
-  border-width: 0;
-  border-bottom: 1px solid var(--color-gray-30);
-  /* splitter grip area */
-  height: 5px;
-  /* make only the splitter border visible */
+#attachment-splitter[location="bottom"][state="collapsed"] {
   margin-top: -5px;
-  /* because of the negative margin needed to make the splitter visible */
-  position: relative;
-  z-index: 10;
-  transition: border-width .3s ease-in;
+  margin-bottom: 0;
 }
 
-#attachment-splitter:not([state="collapsed"]) {
-  border-bottom: 1px solid transparent;
+#attachment-splitter[location="top"]:not([state="collapsed"], :hover, :focus) {
+  border-top-width: 1px;
 }
 
-#attachment-splitter {
-  transition: border-color .3s;
-}
-
-#attachment-splitter[state="collapsed"]:hover {
-  border-bottom: 4px solid var(--selected-item-color);
+/* Collapsed list - splitter on bottom */
+#header-splitter[state="collapsed"] + #attachmentView[location="top"] {
+  border-top-width: 0;
 }
 
 /* ::::: msg header captions ::::: */
 
 .message-header-label {
   padding: 0;
   margin-block: 0;
   margin-inline: 6px 8px;
@@ -483,18 +519,20 @@ button.email-action-button:focus-visible
   position: absolute;
   clip-path: inset(50%);
 }
 
 #attachmentView {
   display: flex;
   flex-direction: column;
   justify-content: stretch;
+  min-height: 30px;
   /* Allow the area to shrink. */
   min-width: 0;
+  max-height: 50%;
 }
 
 #attachmentView,
 #attachmentList {
   border-top: 1px solid var(--splitter-color); /* The same color as the splitters */
 }
 
 :root[lwt-tree] #singleMessage,
